File: Samples/Java/SpringBoot/src/main/java/io/cratis/arc/samples/javaspringboot/TaskRepository.java
// Copyright (c) Cratis. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

package io.cratis.arc.samples.javaspringboot;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Objects;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.Flow;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicReference;
import org.springframework.stereotype.Repository;

/** Thread-safe, in-memory task storage retaining the 100 most recently created tasks. */
@Repository
public final class TaskRepository {
    private static final int MAXIMUM_RETAINED_TASKS = 100;

    private final Object monitor = new Object();
    private final LinkedHashMap<String, StoredTask> tasks = new LinkedHashMap<>();
    private final TaskSnapshotPublisher observableTasks = new TaskSnapshotPublisher(List.of());
    private long stateVersion;

    /** Creates and stores a task, evicting the oldest task when the sample bound is exceeded. */
    public TaskView create(String title) {
        var task = new TaskView(UUID.randomUUID().toString(), title.trim(), false);
        Snapshot committedSnapshot;
        synchronized (monitor) {
            var version = ++stateVersion;
            tasks.put(task.id(), new StoredTask(task, version));
            retainBoundedTasks();
            committedSnapshot = new Snapshot(version, snapshot());
        }
        observableTasks.publish(committedSnapshot);
        return task;
    }

    /** Captures the current task and its stable revision for completion. */
    public TaskCompletionPreparation prepareCompletion(String id) {
        synchronized (monitor) {
            var stored = tasks.get(id);
            return stored == null ? null : new TaskCompletionPreparation(stored.task(), stored.revision());
        }
    }

    /** Captures completion state through a Java asynchronous API. */
    public CompletionStage<TaskCompletionPreparation> prepareCompletionAsync(String id) {
        return CompletableFuture.completedFuture(prepareCompletion(id));
    }

    /** Completes only the exact prepared revision and publishes the committed update. */
    public TaskView complete(TaskCompletionPreparation preparation) {
        TaskView completedTask;
        Snapshot committedSnapshot;
        synchronized (monitor) {
            var current = tasks.get(preparation.task().id());
            if (current == null || current.revision() != preparation.revision()) {
                return null;
            }

            completedTask = new TaskView(current.task().id(), current.task().title(), true);
            var version = ++stateVersion;
            tasks.put(completedTask.id(), new StoredTask(completedTask, version));
            committedSnapshot = new Snapshot(version, snapshot());
        }
        observableTasks.publish(committedSnapshot);
        return completedTask;
    }

    /** Gets a task by identifier. */
    public TaskView byId(String id) {
        synchronized (monitor) {
            var stored = tasks.get(id);
            return stored == null ? null : stored.task();
        }
    }

    /** Gets a task by identifier through a Java asynchronous API. */
    public CompletionStage<TaskView> byIdAsync(String id) {
        return CompletableFuture.completedFuture(byId(id));
    }

    /** Gets a stable snapshot of every retained task ordered by title. */
    public List<TaskView> all() {
        synchronized (monitor) {
            return snapshot();
        }
    }

    /** Observes bounded, replayable snapshots using the JDK Flow API. */
    public Flow.Publisher<List<TaskView>> observe() {
        return observableTasks;
    }

    /** Clears the sample store and publishes the committed empty snapshot. */
    public void clear() {
        Snapshot committedSnapshot;
        synchronized (monitor) {
            tasks.clear();
            committedSnapshot = new Snapshot(++stateVersion, List.of());
        }
        observableTasks.publish(committedSnapshot);
    }

    int subscriberCount() {
        return observableTasks.subscriberCount();
    }

    private List<TaskView> snapshot() {
        var snapshot = tasks.values().stream().map(StoredTask::task).collect(ArrayList<TaskView>::new, List::add, List::addAll);
        snapshot.sort(Comparator.comparing(TaskView::title));
        return List.copyOf(snapshot);
    }

    private void retainBoundedTasks() {
        if (tasks.size() > MAXIMUM_RETAINED_TASKS) {
            tasks.remove(tasks.keySet().iterator().next());
        }
    }

    private record StoredTask(TaskView task, long revision) {
    }

    static final class TaskSnapshotPublisher implements Flow.Publisher<List<TaskView>> {
        private final AtomicReference<Snapshot> current;
        private final CopyOnWriteArrayList<TaskSubscription> subscriptions = new CopyOnWriteArrayList<>();

        TaskSnapshotPublisher(List<TaskView> initial) {
            current = new AtomicReference<>(new Snapshot(0, initial));
        }

        @Override
        public void subscribe(Flow.Subscriber<? super List<TaskView>> subscriber) {
            Objects.requireNonNull(subscriber, "subscriber");
            var subscription = new TaskSubscription(subscriber, subscriptions);
            subscriptions.add(subscription);
            subscription.offer(current.get());
            try {
                subscriber.onSubscribe(subscription);
            } catch (RuntimeException ignored) {
                subscription.cancel();
            }
        }

        void publish(Snapshot snapshot) {
            while (true) {
                var observed = current.get();
                if (snapshot.version() <= observed.version()) {
                    return;
                }
                if (current.compareAndSet(observed, snapshot)) {
                    break;
                }
            }
            subscriptions.forEach(subscription -> subscription.offer(snapshot));
        }

        int subscriberCount() {
            return subscriptions.size();
        }

        Snapshot currentSnapshot() {
            return current.get();
        }
    }

    record Snapshot(long version, List<TaskView> tasks) {
        Snapshot {
            tasks = List.copyOf(tasks);
        }
    }

    static final class TaskSubscription implements Flow.Subscription {
        private static final Object ACTIVE = new Object();
        private static final Object CANCELLED = new Object();
        private static final Object TERMINATED = new Object();

        private final Flow.Subscriber<? super List<TaskView>> subscriber;
        private final CopyOnWriteArrayList<TaskSubscription> owner;
        private final AtomicLong demand = new AtomicLong();
        private final AtomicReference<Snapshot> pending = new AtomicReference<>();
        private final AtomicInteger draining = new AtomicInteger();
        private final AtomicReference<Object> lifecycle = new AtomicReference<>(ACTIVE);
        private final Object offerMonitor = new Object();
        private long latestOfferedVersion = -1;

        TaskSubscription(
            Flow.Subscriber<? super List<TaskView>> subscriber,
            CopyOnWriteArrayList<TaskSubscription> owner) {
            this.subscriber = subscriber;
            this.owner = owner;
        }

        @Override
        public void request(long count) {
            if (count <= 0) {
                fail(new IllegalArgumentException("Flow demand must be positive."));
                return;
            }
            if (lifecycle.get() != ACTIVE) {
                return;
            }
            addDemand(count);
            drain();
        }

        @Override
        public void cancel() {
            if (lifecycle.compareAndSet(ACTIVE, CANCELLED)) {
                pending.set(null);
                owner.remove(this);
            }
        }

        void offer(Snapshot snapshot) {
            synchronized (offerMonitor) {
                if (lifecycle.get() != ACTIVE || snapshot.version() <= latestOfferedVersion) {
                    return;
                }
                latestOfferedVersion = snapshot.version();
                pending.set(snapshot);
            }
            drain();
        }

        private void addDemand(long count) {
            while (true) {
                var existing = demand.get();
                var updated = existing > Long.MAX_VALUE - count ? Long.MAX_VALUE : existing + count;
                if (demand.compareAndSet(existing, updated)) {
                    return;
                }
            }
        }

        private void consumeDemand() {
            while (true) {
                var existing = demand.get();
                if (existing == Long.MAX_VALUE) {
                    return;
                }
                if (demand.compareAndSet(existing, existing - 1)) {
                    return;
                }
            }
        }

        private void drain() {
            if (draining.getAndIncrement() != 0) {
                return;
            }
            var missed = 1;
            do {
                var state = lifecycle.get();
                if (state instanceof Failure failure) {
                    pending.set(null);
                    if (lifecycle.compareAndSet(state, TERMINATED)) {
                        try {
                            subscriber.onError(failure.exception());
                        } catch (RuntimeException ignored) {
                            // Subscriber callbacks cannot compromise repository publication.
                        }
                    }
                } else if (state == ACTIVE && demand.get() > 0) {
                    var snapshot = pending.getAndSet(null);
                    if (snapshot != null && lifecycle.get() == ACTIVE) {
                        consumeDemand();
                        try {
                            subscriber.onNext(snapshot.tasks());
                        } catch (RuntimeException ignored) {
                            cancel();
                        }
                    }
                } else if (state != ACTIVE) {
                    pending.set(null);
                }
                missed = draining.addAndGet(-missed);
            } while (missed != 0);
        }

        private void fail(RuntimeException exception) {
            if (lifecycle.compareAndSet(ACTIVE, new Failure(exception))) {
                pending.set(null);
                owner.remove(this);
                drain();
            }
        }

        private record Failure(RuntimeException exception) {
        }
    }
}

File: Samples/Kotlin/SpringBoot/src/main/kotlin/io/cratis/arc/samples/kotlin/springboot/TaskRepository.kt
// Copyright (c) Cratis. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

package io.cratis.arc.samples.kotlin.springboot

import java.util.UUID
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Profile

/**
 * Everything the task board needs from a store.
 *
 * Commands and queries depend on this interface, never on a concrete store. That is what lets
 * `./run.sh --database mongodb` swap the whole persistence layer without a single edit to
 * [CreateTask], [CompleteTask] or [TaskView] — the artifacts describe intent, and where the state
 * lives is a separate decision.
 */
public interface TaskRepository {
    /** Creates and stores a task. */
    public fun create(title: String): TaskView

    /** Captures the current task and its stable revision for completion. */
    public fun prepareCompletion(id: String): TaskCompletionPreparation?

    /** Completes only the exact prepared revision and publishes the committed update. */
    public fun complete(preparation: TaskCompletionPreparation): TaskView?

    /** Gets a task by identifier. */
    public fun byId(id: String): TaskView?

    /** Gets a stable snapshot of every retained task ordered by title. */
    public fun all(): List<TaskView>

    /** Observes stable snapshots of every retained task. */
    public fun observe(): Flow<List<TaskView>>

    /** Clears the sample store. */
    public fun clear()
}

/** Thread-safe, in-memory task storage retaining the 100 most recently created tasks. */
public class InMemoryTaskRepository : TaskRepository {
    private val monitor: Any = Any()
    private val tasks: LinkedHashMap<String, StoredTask> = LinkedHashMap()
    private val observableTasks: MutableStateFlow<List<TaskView>> = MutableStateFlow(emptyList())
    private var nextRevision: Long = 0

    /** Creates and stores a task, evicting the oldest task when the fixed sample bound is exceeded. */
    override fun create(title: String): TaskView {
        val task = TaskView(UUID.randomUUID().toString(), title.trim(), completed = false)
        synchronized(monitor) {
            tasks[task.id] = StoredTask(task, ++nextRevision)
            retainBoundedTasks()
            observableTasks.value = snapshot()
        }
        return task
    }

    override fun prepareCompletion(id: String): TaskCompletionPreparation? = synchronized(monitor) {
        tasks[id]?.let { stored -> TaskCompletionPreparation(stored.task, stored.revision) }
    }

    override fun complete(preparation: TaskCompletionPreparation): TaskView? = synchronized(monitor) {
        val current = tasks[preparation.task.id]
        if (current == null || current.revision != preparation.revision) {
            return@synchronized null
        }

        val completedTask = current.task.copy(completed = true)
        tasks[completedTask.id] = StoredTask(completedTask, ++nextRevision)
        observableTasks.value = snapshot()
        completedTask
    }

    override fun byId(id: String): TaskView? = synchronized(monitor) { tasks[id]?.task }

    override fun all(): List<TaskView> = synchronized(monitor) { snapshot() }

    override fun observe(): Flow<List<TaskView>> = observableTasks

    override fun clear() {
        synchronized(monitor) {
            tasks.clear()
            observableTasks.value = emptyList()
        }
    }

    private fun snapshot(): List<TaskView> = tasks.values.map(StoredTask::task).sortedBy(TaskView::title)

    private fun retainBoundedTasks() {
        if (tasks.size > MAXIMUM_RETAINED_TASKS) {
            tasks.remove(tasks.keys.first())
        }
    }

    private data class StoredTask(val task: TaskView, val revision: Long)

    private companion object {
        const val MAXIMUM_RETAINED_TASKS: Int = 100
    }
}

/**
 * Supplies the in-memory store unless a database profile selected another one.
 *
 * The condition is a profile expression rather than `@ConditionalOnMissingBean`, which is only
 * reliable inside an auto-configuration: in an application configuration it is evaluated in
 * bean-definition order, so whether it sees the database store depends on component-scan order.
 * That is exactly the kind of "works on my machine" a sample should not teach.
 */
@Configuration(proxyBeanMethods = false)
@Profile("!mongodb & !postgres")
public class InMemoryTaskRepositoryConfiguration {
    /** The default store: no container, no connection string, no setup. */
    @Bean
    public fun inMemoryTaskRepository(): TaskRepository = InMemoryTaskRepository()
}
